File: app/src/main/java/com/elitec/alejotaller/feature/product/domain/entity/Product.kt
package com.elitec.alejotaller.feature.product.domain.entity

data class Product(
    val id: String,
    val name: String,
    val description: String,
    val price: Double,
    val photoUrl: String,
    val categoryId: String,
    val rating: Double = 0.0,
    val photoLocalResource: Int? = null,
    /** Unidades físicas en almacén (no descontadas hasta VERIFIED). */
    val existence: Int = 0,
    /** Unidades comprometidas en pedidos UNVERIFIED (soft-hold). */
    val reserved: Int = 0
) {
    init {
        require(id != "") { "The value of product identifier cant not by empty" }
        require(price >= 0.0) { "The price of product identifier cant not by a negative" }
        require(existence >= 0) { "Product existence (stock) cannot be negative" }
        require(reserved >= 0) { "Product reserved stock cannot be negative" }
    }

    /** Stock vendible ahora = existence - reserved */
    fun availableStock(): Int = (existence - reserved).coerceAtLeast(0)
}

File: web/src/core/feature/product/domain/entity/Product.ts
export interface Product {
    id: string
    name: string
    description: string
    /** Unidades físicas en almacén (no descontadas hasta VERIFIED) */
    existence: number
    /** Unidades comprometidas en pedidos UNVERIFIED (soft-hold) */
    reserved: number
    price: number
    photoUrl: string
    categoryId: string
    rating?: number
    photoLocalResource?: number | null
    createdAtIso?: string
}

/** Stock vendible ahora = existence - reserved */
export function availableStock(product: Product): number {
    const existence = Math.max(0, Math.floor(product.existence ?? 0))
    const reserved = Math.max(0, Math.floor(product.reserved ?? 0))
    return Math.max(0, existence - reserved)
}

/**
 * Factory con validaciones (equivalente al init{})
 */
export function createProduct(product: Product): Product {
    if (!product.id || product.id.trim() === "") {
        throw new Error("The value of product identifier cannot be empty")
    }

    if (product.price < 0) {
        throw new Error("The price cannot be negative")
    }

    if (product.existence == null || Number.isNaN(product.existence) || product.existence < 0) {
        throw new Error("Product existence (stock) cannot be negative")
    }

    const reserved = product.reserved == null || Number.isNaN(product.reserved)
        ? 0
        : Math.floor(product.reserved)

    if (reserved < 0) {
        throw new Error("Product reserved stock cannot be negative")
    }

    return {
        rating: 0.0,
        ...product,
        existence: Math.floor(product.existence),
        reserved
    }
}
